scheduling/taskmetadata.go

```diff
@@ -0,0 +1,463 @@
+package scheduling
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/featureform/locker"
+	ss "github.com/featureform/storage_storer"
+)
+
+type TaskMetadataList []TaskMetadata
+
+func (tml *TaskMetadataList) ToJSON() string {
+	return ""
+}
+
+type TaskRunList []TaskRunMetadata
+
+func (trl *TaskRunList) ToJSON() string {
+	return ""
+}
+
+func (trl *TaskRunList) FilterByStatus(status Status) {
+	var newList TaskRunList
+	for _, run := range *trl {
+		if run.Status == status {
+			newList = append(newList, run)
+		}
+	}
+	*trl = newList
+}
+
+type TaskMetadataManager struct {
+	// storage ss.MetadataStorer
+	storer ss.MetadataStorer
+}
+
+func NewMemoryTaskMetadataManager() TaskMetadataManager {
+	memoryLocker := locker.MemoryLocker{
+		LockedItems: sync.Map{},
+		Mutex:       &sync.Mutex{},
+	}
+
+	memoryStorer := ss.MemoryStorerImplementation{
+		Storage: make(map[string]string),
+	}
+
+	memoryMetadataStorer := ss.MetadataStorer{
+		Locker: &memoryLocker,
+		Storer: &memoryStorer,
+	}
+	return TaskMetadataManager{
+		storer: memoryMetadataStorer,
+	}
+}
+
+func (m *TaskMetadataManager) CreateTask(name string, tType TaskType, target TaskTarget) (TaskMetadata, error) {
+	keys, err := m.storer.List(TaskMetadataKey{}.String())
+	if err != nil {
+		return TaskMetadata{}, fmt.Errorf("failed to fetch keys: %v", err)
+	}
+
+	// This logic could probably be somewhere else
+	var latestID int
+	if len(keys) == 0 {
+		latestID = 0
+	} else {
+		latestID, err = getLatestId(keys)
+		if err != nil {
+			return TaskMetadata{}, err
+		}
+	}
+
+	metadata := TaskMetadata{
+		ID:          TaskID(latestID + 1),
+		Name:        name,
+		TaskType:    tType,
+		Target:      target,
+		TargetType:  target.Type(),
+		DateCreated: time.Now().UTC(),
+	}
+
+	// I do this serialize and deserialize a lot in this file. Would be nice to have set and get helpers that deal with
+	// all the converting instead
+	serializedMetadata, err := metadata.Marshal()
+	if err != nil {
+		return TaskMetadata{}, fmt.Errorf("failed to marshal metadata: %v", err)
+	}
+
+	key := TaskMetadataKey{taskID: metadata.ID}
+	err = m.storer.Create(key.String(), string(serializedMetadata))
+	if err != nil {
+		return TaskMetadata{}, fmt.Errorf("failed to create task metadata: %v", err)
+	}
+
+	runs := TaskRuns{
+		TaskID: metadata.ID,
+		Runs:   []TaskRunSimple{},
+	}
+	serializedRuns, err := runs.Marshal()
+	if err != nil {
+		return TaskMetadata{}, err
```

Comment:
should return an `fmt.Errorf` for clarity

Reply:
Marshal returns a featureform error